manager/container_test.go

```diff
@@ -226,3 +231,69 @@ func TestUpdateNvidiaStats(t *testing.T) {
 	assert.NotNil(t, err)
 	assert.Equal(t, info.ContainerStats{}, stats)
 }
+
+func TestOnDemandHousekeeping(t *testing.T) {
+	statsList := itest.GenerateRandomStats(1, 4, 1*time.Second)
+	stats := statsList[0]
+
+	cd, mockHandler, memoryCache, fakeClock := newTestContainerData(t)
+	mockHandler.On("GetStats").Return(stats, nil)
+	defer cd.Stop()
+
+	// 0 seconds should always trigger an update
+	go cd.OnDemandHousekeeping(0 * time.Second)
+	cd.housekeepingTick(fakeClock.NewTimer(time.Minute).C(), testLongHousekeeping)
+
+	fakeClock.Step(2 * time.Second)
+
+	// This should return without requiring a housekeepingTick because stats have been updated recently enough
+	cd.OnDemandHousekeeping(3 * time.Second)
+
+	go cd.OnDemandHousekeeping(1 * time.Second)
+	cd.housekeepingTick(fakeClock.NewTimer(time.Minute).C(), testLongHousekeeping)
+
+	checkNumStats(t, memoryCache, 2)
+	mockHandler.AssertExpectations(t)
+}
+
+func TestConcurrentOnDemandHousekeeping(t *testing.T) {
+	statsList := itest.GenerateRandomStats(1, 4, 1*time.Second)
+	stats := statsList[0]
+
+	cd, mockHandler, memoryCache, fakeClock := newTestContainerData(t)
+	mockHandler.On("GetStats").Return(stats, nil)
+	defer cd.Stop()
+
+	var waitForQueue, waitForHousekeeping sync.WaitGroup
+	// This is a copy of the OnDemandHousekeeping function with sync.WaitGroup plummed
+	// through so that progress in each function can be synchronized to ensure that they
+	// are progressing concurrently.
+	mockedOnDemandHousekeeping := func(c *containerData, maxAge time.Duration) {
+		if c.clock.Since(c.statsLastUpdatedTime) > maxAge {
+			housekeepingFinishedChan := make(chan struct{})
+			c.onDemandChan <- housekeepingFinishedChan
+			waitForQueue.Done()
+			select {
+			case <-c.stop:
+			case <-housekeepingFinishedChan:
+			}
+			waitForHousekeeping.Done()
+		}
+	}
```

Comment:
Correct me if I'm wrong, but I think you can get away with not mocking this:

```go
onDemandCache := make(chan chan struct{}, numConcurrentCalls)
for i := 0; i < numConcurrentCalls; i++ {
	go func() {
		c.OnDemandHousekeeping(0*time.Second)
		waitForHousekeeping.Done()
	}
	// Wait for work to be queued
	onDemandCache = append(onDemandCache, <-c.onDemandChan)
}
// Requeue work:
for _, ch := range onDemandCache {
	c.OnDemandChan <- ch
}

Reply:
I can switch to that.

Reply:
done